mapper/pojo-base/src/main/java/org/hibernate/search/mapper/pojo/massindexing/impl/PojoMassIndexingBatchCoordinator.java

```diff
@@ -174,17 +179,21 @@ protected void cleanUpOnInterruption() throws InterruptedException {
 			// Indexing performed before the exception must still be committed,
 			// in order to leave the index in a consistent state
 			closer.push( PojoMassIndexingBatchCoordinator::flushAndRefresh, this );
-			closer.push( PojoMassIndexerAgent::stop, agent );
-			agent = null;
+			for ( BatchCoordinatorContext context : coordinatorContexts ) {
+				closer.push( PojoMassIndexerAgent::stop, context.agent() );
+				context.withAgent( null );
+			}
 		}
 	}
 
 	@Override
 	protected void cleanUpOnFailure() {
 		try ( Closer<RuntimeException> closer = new Closer<>() ) {
 			closer.pushAll( this::cancelPendingTask, indexingFutures );
-			closer.push( PojoMassIndexerAgent::stop, agent );
-			agent = null;
+			for ( BatchCoordinatorContext context : coordinatorContexts ) {
+				closer.push( PojoMassIndexerAgent::stop, context.agent() );
+				context.withAgent( null );
+			}
```

Comment:
If you follow my suggestion to create contexts within `PojoMasIndexingBatchCoordinator`, you don't necessarily need a loop, you can do this:

```suggestion
			closer.pushAll( PojoMassIndexerAgent::stop, coordinatorContexts, BatchCoordinatorContext::agent );
			coordinatorContexts.clear();
```